src/hardware/signal_generator.rs

```diff
@@ -0,0 +1,219 @@
+    }
+
+    // Increment the phase of the signal.
+    //
+    // # Note
+    // This handles automatically applying pending configurations on phase wrap.
+    //
+    // # Returns
+    // The new phase to use
+    fn increment(&mut self) -> i32 {
+        let (phase, overflow) = self
+            .phase_accumulator
+            .overflowing_add(self.config.frequency_tuning_word);
+
+        self.phase_accumulator = phase;
+
+        // Special case: If the FTW is specified as zero, we would otherwise never update the
+        // settings. Perform a check here for this corner case.
+        if overflow || self.config.frequency_tuning_word == 0 {
+            if let Some(config) = self.pending_config.take() {
+                self.config = config;
+                self.phase_accumulator = 0;
+            }
+        }
+
+        self.phase_accumulator as i32
+    }
+
+    /// Skip `count` elements of the generator
+    pub fn skip(&mut self, count: u32) {
+        for _ in 0..count {
+            self.increment();
+        }
+    }
+
+    /// Get the next value in the generator sequence.
+    pub fn next(&mut self) -> i16 {
+        let phase = self.increment();
+
+        let amplitude = match self.config.signal {
+            Signal::Sine => (dsp::cossin(phase).1 >> 16) as i16,
+            Signal::Square => {
+                if phase < self.config.phase_symmetry {
+                    i16::MAX
+                } else {
+                    i16::MIN
+                }
+            }
+            Signal::Triangle => {
+                if phase < self.config.phase_symmetry {
+                    let duration_of_phase =
+                        (self.config.phase_symmetry.wrapping_sub(i32::MIN)
+                            >> 16) as u16;
+                    let phase_progress =
+                        (phase.wrapping_sub(i32::MIN) >> 16) as u16;
+
+                    if duration_of_phase == 0 {
+                        i16::MIN
+                    } else {
+                        i16::MIN.wrapping_add(
+                            (u16::MAX as u32 * phase_progress as u32
+                                / duration_of_phase as u32)
+                                as i16,
+                        )
+                    }
+                } else {
+                    let duration_of_phase =
+                        (i32::MAX.wrapping_sub(self.config.phase_symmetry)
+                            >> 16) as u16;
+                    let phase_progress = (phase
+                        .wrapping_sub(self.config.phase_symmetry)
+                        >> 16) as u16;
+
+                    if duration_of_phase == 0 {
+                        i16::MAX
+                    } else {
+                        i16::MAX.wrapping_sub(
+                            (u16::MAX as u32 * phase_progress as u32
+                                / duration_of_phase as u32)
+                                as i16,
+                        )
+                    }
```

Comment:
1. It's not immediately clear to me how we could combine the code paths given that we are calculating an asymmetric triangle wave. I'm not aware of a general equation, and the only approach I'm aware of is to treat it like a PWL function, which is why there is a branch in code paths here (one for the rising side, one for the falling).
2. Ultimately, if we're mapping a given phase to a value in the triangular wave, we have to know the ratio of our current phase to the duration of the current phase. While restricting asymmetry to a power-of-two would simplify calculations for one phase (e.g. rising), the inverse phase duration would then _not_ be a power of two (by definition) of the phase, so the calculation can't be simplified on that end.

Open to some concrete suggestions otherwise here though.

Reply:
It helps to clean up terminology. The phase in `duration_of_phase` is not the phase but probably more "segment".
Store your `1/duration_of_phase` for the two segments in the configuration. It only needs to be computed once. It's just a frequency tuning word again.

Reply:
This boils down to two separate frequency tuning words (one for each segment). That looks like the easy solution and can equally be used for square. For cosine the weird sidefeect of supporting an asymmetric cosine if also fine.

Reply:
I worry about using an "amplitude tuning word" as I think it might result in the waveform being ill-formed. GIven that we use a FTW, there's no guarantee that we have exactly an integer number of steps in each phase, but the amplitude tuning word would assume as such.

In that case, I imagine the triangle wave may "drift" up and down depending on the number of tuning word occurrences in each phase.

Reply:
I think this is resolved by redefining `duration_of_phase` to be constant values calculated at config time. I still need to test the updated implementation though.

Reply:
No "amplitude tuning word", just two different FTWs, one for `phase >= 0` and the other for `phase < 0` (phase being `i32`). That will not drift and is completely accurate.

Reply:
```rust
phase = phase.wrapping_add(if phase >= 0 {
    config.frequency_tuning_word[0]
} else {
    config.frequency_tuning_word[1]
});
```

for all waveforms. Square output is `(2*sign bit - 1)*amplitude` and triangle output is `(phase.abs() >> 16)*amplitude >> 16`. The user visible tuning word is then the weighted (with the asymmetry and 1-asymmetry respectively) harmonic mean of the two (up and down) tuning words.

Reply:
This method mentioned above results in a _much_ simpler implementaiton that should be much faster as well.